README.md
# react-native-confirmation-code-field

[![npm](https://img.shields.io/npm/v/react-native-confirmation-code-field.svg)](https://www.npmjs.com/package/react-native-confirmation-code-field)
[![Travis](https://img.shields.io/travis/retyui/react-native-confirmation-code-field.svg?label=unix)](https://travis-ci.org/retyui/react-native-confirmation-code-field)

A react-native confirmation code field compatible with iOS, Android and Web Platforms (based on [this](https://github.com/ttdung11t2/react-native-confirmation-code-input) project [Migration Guide](docs/migration.md))

### Component features:

- 🔮 Simple. Easy to use;
- 🍎 Support "fast paste SMS-code" on iOS. And custom code paste for Android;
- 🚮 Clearing part of the code by clicking on the cell;
- ⚡ `blur()` and `focus()` methods;
- 🛠 Extendable and hackable;
- 🤓 Readable [changelog](CHANGELOG.md).

## Links

- [API documentation](docs/API.md)
- [Examples](examples/src/realDemo)
- Live demos [iOS / Android](https://snack.expo.io/@retyui/demo-for-react-native-confirmation-code-field), [react-native-web](https://react-native-confirmation-code-field.netlify.com/)

## Screenshots

<a href="https://github.com/retyui/react-native-confirmation-code-field/tree/master/examples/src/realDemo/AnimatedExample"><img width="250" src="https://raw.githubusercontent.com/retyui/react-native-confirmation-code-field/master/docs/img/animated.gif"/></a><a href="https://github.com/retyui/react-native-confirmation-code-field/tree/master/examples/src/realDemo/RedExample"><img width="250" src="https://raw.githubusercontent.com/retyui/react-native-confirmation-code-field/master/docs/img/red.gif"/></a><a href="https://github.com/retyui/react-native-confirmation-code-field/tree/master/examples/src/realDemo/DarkExample"><img width="250" src="https://raw.githubusercontent.com/retyui/react-native-confirmation-code-field/master/docs/img/dark.gif"/></a>

## Install

```sh
yarn add react-native-confirmation-code-field
# or
npm install react-native-confirmation-code-field
```

## Usage

```js
import React, { Component } from 'react';
import CodeInput from 'react-native-confirmation-code-field';

class App extends Component {
  handlerOnFulfill = code => console.log(code);

  render() {
    return <CodeInput onFulfill={this.handlerOnFulfill} />;
  }
}
```

## How paste or clear code

Paste code can helpful for Android platform when you can read SMS.

```js
import React, { Component, createRef } from 'react';
import CodeInput from 'react-native-confirmation-code-field';

class App extends Component {
  handlerOnFulfill = code => {
    if (isValidCode(code)) {
      console.log(code);
    } else {
      this.clearCode();
    }
  };

  field = createRef();

  clearCode() {
    const { current } = this.field;

    if (current) {
      current.clear();
    }
  }

  pasteCode() {
    const { current } = this.field;

    if (current) {
      current.handlerOnTextChange(value);
    }
  }

  render() {
    return <CodeInput ref={this.field} onFulfill={this.handlerOnFulfill} />;
  }
}
```

## Analogs

- [react-native-keycode](https://github.com/leanmotherfuckers/react-native-keycode)
- [react-native-pin-code](https://github.com/gkueny/react-native-pin-code)

## How it works?

This component consists of:

1. Container `<View {...containerProps}/>`;
2. Render the "Cells" for the text code inside the container ("Cells" is `<TextInput {...cellProps} />`);
3. And over this render invisible `<TextInput {...inputProps}/>`;
4. "Cursor" inside cell is [simulated component](src/components/Cursor.js)


CHANGELOG.md
# Change Log

This project adheres to [Semantic Versioning](http://semver.org/).

## 3.6.0

- Add new prop `blurOnSubmit?: boolean` [#62](https://github.com/retyui/react-native-confirmation-code-field/pull/62) by [leozzitowned](https://github.com/leozzitowned)

## 3.5.0

- Add new prop `CellComponent?: ComponentType`, it can be useful for create some animations [Example](https://github.com/retyui/react-native-confirmation-code-field/tree/master/examples/src/realDemo/AnimatedExample)

## 3.4.1

- Add RTL support [#55](https://github.com/retyui/react-native-confirmation-code-field/pull/55) by [bamlhs](https://github.com/bamlhs)

## 3.4.0

- Added support for `react-native-web` [#50](https://github.com/retyui/react-native-confirmation-code-field/issues/50)

## 3.3.0

- Added new prop `normalizeCode: (code: string) => string` via [#47](https://github.com/retyui/react-native-confirmation-code-field/issues/47) [#41](https://github.com/retyui/react-native-confirmation-code-field/pull/41)

## 3.2.3

- Fix onPress event handler [#45](https://github.com/retyui/react-native-confirmation-code-field/issues/45)

## 3.2.2

- Update typings [#43 index.d.ts](https://github.com/retyui/react-native-confirmation-code-field/pull/43/files#diff-b52768974e6bc0faccb7d4b75b162c99)
- Add example how work with `paste()` and `clean()` to README.md [#43 README.md](https://github.com/retyui/react-native-confirmation-code-field/pull/43/files#diff-04c6e90faac2675aa89e2176d2eec7d8)
- Recreate example for RN@0.59
- Update `devDependencies`

## 3.2.1

- Fix issue about `testID` [#38](https://github.com/retyui/react-native-confirmation-code-field/issues/38)

## 3.2.0

- Fix Flow.js error for support `react-native@0.58` [#33](https://github.com/retyui/react-native-confirmation-code-field/pull/33)

## 3.1.3

- Fix `autoFucus` prop [#32](https://github.com/retyui/react-native-confirmation-code-field/pull/32)

## 3.1.0

- Fix iOS borderBottom style [#28](https://github.com/retyui/react-native-confirmation-code-field/pull/28/files)

## 3.0.0

- Remove base implementation based on `One code cell` === `One TextInput`
- Fix iOS fast paste SMS code
- Update `inputProps: Function` prop, now is `inputProps: ?Object`
- Remove `canPasteCode`, now it is works always by default
- Remove `inputStyle: Function` use `cellProps`
- Remove `onChangeText` use `inputProps={{onChangeText: (text) => {}}}`
- Remove some `variant` values: `border-b-t` and `border-l-r`
- Add new component `focus()` and `blur()` methods
- Fix typescript definition [#27](https://github.com/retyui/react-native-confirmation-code-field/pull/27/files)

## 2.0.5

- Fix typescript definition [#27](https://github.com/retyui/react-native-confirmation-code-field/pull/27/files)

## 2.0.4

- Fix Flow.js types
- Fix spread non-iterable instance [#24](https://github.com/retyui/react-native-confirmation-code-field/pull/24)

## 2.0.1

- Update TypeScript definition [changes](https://github.com/retyui/react-native-confirmation-code-field/commit/5e5bf0b9f912dc49bbfb5c7a24c78004374b089f#diff-b52768974e6bc0faccb7d4b75b162c99)

## 2.0.0

- New example for RN 0.57 version
- Merged changes from `1.2.1` version
- Update dependencies
  > RN@0.57 - Fixed extreme `<TextInput>` slowness ([5017b86](https://github.com/facebook/react-native/commit/5017b86) by [@gnprice](https://github.com/gnprice))
- Remove `ignoreCaseWhenCompareCode` and `compareWithCode` props.
- Add paste support. By default it's disabled. That would enable set `canPasteCode={true}`
- Rename property `getInputProps` => `inputProps`
- Rename property `getInputStyle` => `inputStyle`
- Remove iOS hack for `onKeyPress` (use latest RN version)
- Remove `INDEX` typing
- Fix desynchronization `onFocus` and `setState`
  [#16](https://github.com/retyui/react-native-confirmation-code-field/issues/16)

## 1.2.3

- Fix desynchronization `onFocus` and `setState`
  [#16](https://github.com/retyui/react-native-confirmation-code-field/issues/16)

## 1.2.1

- Add export for types [#10](https://github.com/retyui/react-native-confirmation-code-field/pull/10)

## 1.2.0

- Add new prop `maskSymbol?: string` [#6](https://github.com/retyui/react-native-confirmation-code-field/issues/6)

## 1.1.0

- Add `keyboardType` prop [#4](https://github.com/retyui/react-native-confirmation-code-field/pull/4)
- Fix typescript module name [#3](https://github.com/retyui/react-native-confirmation-code-field/pull/3)

## 1.0.0

- Rename `ignoreCase` => `ignoreCaseWhenCompareCode` property
- Rename `className` => `variant` property

- Add new methods (`getInputStyle()`and `getInputProps()`\_ for help users to customize inputs
- Add new property `containerProps`, `testID`
- Add `onChangeCode()` callback and `defaultCode` props. [#22](https://github.com/ttdung11t2/react-native-confirmation-code-input/pull/22)[#33](https://github.com/ttdung11t2/react-native-confirmation-code-input/pull/33/files)
- Add `clear()` method to TypeScript definition [#21](https://github.com/ttdung11t2/react-native-confirmation-code-input/pull/21)

- Fix slowly focus change [10](https://github.com/ttdung11t2/react-native-confirmation-code-input/pull/10)
- Fix iOS input bag. [#38](https://github.com/ttdung11t2/react-native-confirmation-code-input/pull/38/files)

# 0.0.0

- Fork dead repository


examples/src/realDemo/RedExample/styles.js
import { StyleSheet } from 'react-native';

export default StyleSheet.create({
  linearGradient: {
    minHeight: 555,
  },
  inputWrapper: {
    alignItems: 'center',
    justifyContent: 'center',
  },
  inputLabel: {
    paddingTop: 100,
    paddingBottom: 10,
    color: '#fff',
    fontSize: 25,
    fontWeight: '700',
    textAlign: 'center',
  },
  inputSubLabel: {
    color: '#fff',
  },
  inputWrapStyle: {
    height: 50,
    marginTop: 30,
  },
  input: {
    height: 50,
    width: 40,
    borderRadius: 3,
    color: '#fff',
    backgroundColor: 'rgba(255,255,255,0.3)',
  },
  inputNotEmpty: {
    backgroundColor: 'rgba(0,0,0,0)',
  },
  nextButton: {
    marginTop: 100,
    width: 70,
    height: 70,
    borderRadius: 80,
    backgroundColor: '#fff',
    alignItems: 'center',
    justifyContent: 'center',
    // IOS
    shadowColor: '#000',
    shadowOffset: { width: 0, height: 2 },
    shadowOpacity: 0.2,
    // Android
    elevation: 5,
  },
  nextButtonArrow: {
    transform: [{ translateX: -3 }, { rotate: '45deg' }],
    borderColor: '#ff595f',
    width: 20,
    height: 20,
    borderWidth: 4,
    borderLeftWidth: 0,
    borderBottomWidth: 0,
  },
});


examples/src/realDemo/AnimatedExample/index.js
import React, { Component } from 'react';
import { Alert, Animated, Image, Text, View } from 'react-native';

import CodeFiled from 'react-native-confirmation-code-field';

import styles, {
  ACTIVE_CELL_BG_COLOR,
  CELL_BORDER_RADIUS,
  CELL_SIZE,
  DEFAULT_CELL_BG_COLOR,
  NOT_EMPTY_CELL_BG_COLOR,
} from './styles';

const codeLength = 4;

const source = {
  uri:
    'https://user-images.githubusercontent.com/4661784/56352614-4631a680-61d8-11e9-880d-86ecb053413d.png',
};

export default class AnimatedExample extends Component {
  _animationsColor = [...new Array(codeLength)].map(
    () => new Animated.Value(0),
  );
  _animationsScale = [...new Array(codeLength)].map(
    () => new Animated.Value(1),
  );

  onFinishCheckingCode = code => {
    if (code !== '1234') {
      return Alert.alert(
        'Confirmation Code',
        'Code not match! Try 1234',
        [{ text: 'OK' }],
        { cancelable: false },
      );
    }

    Alert.alert('Confirmation Code', 'Successful!', [{ text: 'OK' }], {
      cancelable: false,
    });
  };

  animateCell({ hasValue, index, isFocused }) {
    Animated.parallel([
      Animated.timing(this._animationsColor[index], {
        toValue: isFocused ? 1 : 0,
        duration: 250,
      }),
      Animated.spring(this._animationsScale[index], {
        toValue: hasValue ? 0 : 1,
        duration: hasValue ? 300 : 250,
      }),
    ]).start();
  }

  cellProps = ({ hasValue, index, isFocused }) => {
    const animatedCellStyle = {
      backgroundColor: hasValue
        ? this._animationsScale[index].interpolate({
            inputRange: [0, 1],
            outputRange: [NOT_EMPTY_CELL_BG_COLOR, ACTIVE_CELL_BG_COLOR],
          })
        : this._animationsColor[index].interpolate({
            inputRange: [0, 1],
            outputRange: [DEFAULT_CELL_BG_COLOR, ACTIVE_CELL_BG_COLOR],
          }),
      borderRadius: this._animationsScale[index].interpolate({
        inputRange: [0, 1],
        outputRange: [CELL_SIZE, CELL_BORDER_RADIUS],
      }),
      transform: [
        {
          scale: this._animationsScale[index].interpolate({
            inputRange: [0, 1],
            outputRange: [0.2, 1],
          }),
        },
      ],
    };

    // Run animation on next event loop tik
    // Because we need first return new style prop and then animate this value
    setTimeout(() => {
      this.animateCell({ hasValue, index, isFocused });
    }, 0);

    return {
      style: [styles.input, animatedCellStyle],
    };
  };

  containerProps = { style: styles.inputWrapStyle };

  render() {
    /*concept : https://dribbble.com/shots/5476562-Forgot-Password-Verification/attachments */
    return (
      <View style={styles.inputWrapper}>
        <Text style={styles.inputLabel}>Verification</Text>
        <Image style={styles.icon} source={source} />
        <Text style={styles.inputSubLabel}>
          {'Please enter the verification code\nwe send to your email address'}
        </Text>
        <CodeFiled
          maskSymbol=" "
          variant="clear"
          codeLength={codeLength}
          keyboardType="numeric"
          cellProps={this.cellProps}
          containerProps={this.containerProps}
          onFulfill={this.onFinishCheckingCode}
          CellComponent={Animated.Text}
        />
        <View style={styles.nextButton}>
          <Text style={styles.nextButtonText}>Verify</Text>
        </View>
      </View>
    );
  }
}
